### ISSUE TITLE
covert darknet weight to darkflow pb file detect no object

### ISSUE THREAD

**Issue by liuxiaozhe** [thtrieu/darkflow#1003]
i use darknet to train  and the weight ,  but when trans  darknet weight to darkflow pb  weight file, no object can be detected

**Comment by skofot** [comment]
Did you use pre-trained weights? I have no experience in darknet, but I know that when training a model in darkflow without pre-trained weights the .pb-output is unusable after ~14000 steps. See #330. Thinking that there might be a similar issue even if you've trained in darknet.

**Comment by dreistheman** [comment]
@skofot I'm currently training at about 13000 steps now with tiny-yolo-voc as base. So it's safe to train beyond 14000 in my case?

**Comment by skofot** [comment]
@dreistheman As long as you've used the tiny-yolo-voc pre-trained weights when beginning your training there should be no worries, at least considering this particular issue.

**Comment by dreistheman** [comment]
@skofot Thank you for replying, it's nice to hear that. :) I have 2 small questions more if it's okay for you sir. 
First is: I'm training the model for my 640x480 video dataset but just discovered the width and height param in the cfg which are at 416x416. How woud that affect the performance when tested with larger resolution images like the said dataset?

Secondly, is the _--epoch 10000_ parameter in the darkflow command viable? I followed it from this tutorial: https://sites.google.com/view/tensorflow-example-java-api/complete-guide-to-train-yolo/train-yolo-with-darkflow

**Comment by Wanderish** [comment]
@skofot I have trained **initially using the tiny-yolo-voc.weights** for my custom classes for up to 11250 steps.
I was using my model in a server and was trying to load the weights from .pb file. Everything is happening, no error whatsoever but there are no predictions of any kind.
What do you think might be the problem? Any suggestions?

**Comment by mpky** [comment]
@Wanderish did you try turning the threshold down? Sometimes the prediction confidence is quite low

**Comment by skofot** [comment]
@Wanderish 11250 steps is not too many steps, so you might need to give it more time. As stated above you could try lowering the threshold to see if anything is ever detected. Also, did you train using `--trainer adam`? Any time I've trained a model using the default optimizer I've had problems getting any output. There is also the possibility of your training set being insufficient. 

Do you get any output when loading and testing the checkpoint generated while training? I'm thinking if the .pb-file is causing the error or if it's the model itself.

**Comment by mpky** [comment]
@skofot Is the default optimizer not Adam?? Oof

**Comment by skofot** [comment]
@mpky I don't believe it is. I trained a model without `--trainer adam` specifically to compare them and it behaved very differently, giving a training loss of numbers around ~E-6 (compared to ~1 when I normally train) but no output.

**Comment by mpky** [comment]
@skofot good to know. any idea what the default is? SGD?

**Comment by skofot** [comment]
@mpky I looked at the defaults.py file and it looks like the optimizer is RMSprop by the code below.

```python
def setDefaults(self):
    self.define('trainer', 'rmsprop', 'training algorithm')

**Comment by Wanderish** [comment]
> @Wanderish did you try turning the threshold down? Sometimes the prediction confidence is quite low

@mpky Yes, I did. That actually did the trick but the prediction was sometimes off the track. It was swap predicting but after testing for 10-20 times, it was predicting well.

Also,

> @Wanderish 11250 steps is not too many steps, so you might need to give it more time. As stated above you could try lowering the threshold to see if anything is ever detected. Also, did you train using `--trainer adam`? Any time I've trained a model without it I've had problems getting any output. There is also the possibility of your training set being insufficient.
> 
> Do you get any output when loading and testing the checkpoint generated while training? I'm thinking if the .pb-file is causing the error or if it's the model itself.

@skofot no I did not use adam optimizer. Is it really necessary? I mean I have no knowledge as to why it is used. Could you shed some light on it? And you are actually right about the training set being insufficient. I used only about 300 images/class but I'm training again in a few days, collecting ~2k images now/class. Thanks for your thoughts :)

**Comment by skofot** [comment]
@Wanderish I don't know too much about optimizers more than that their purpose is to make training more efficient, and that the best optimizer will vary between datasets. When I didn't use the Adam optimizer I had problems getting any output at all. But since you got output when lowering the threshold the default seems to work fine and the model is just not ready. You could keep training your model with the same data for some tens of thousands of steps more and it might make you happy. Even if the loss stops decreasing the model is still learning.

If you want to create the perfect model, you could train new models using different optimizers to the same amount of steps and use the same data to see if the optimizers have any major impact for you.

**Comment by Wanderish** [comment]
@skofot I would've trained it for a longer time but I had to get the model ready for my project demonstration. So I had to cut it out. Also I've been training using CPU i5 7th gen even though the system has an NVIDIA 1050Ti 4GB graphic card. There's some dependencies which need to be installed for that. Got stuck, so thought I'd just run it on CPU for the demo. 
Trying with the GPU now. Hoping to get some results! :)

Thanks for the advice!

**Comment by skofot** [comment]
@Wanderish Good luck setting it up! It's a whole new world of speed enabling GPU training.

**Comment by Wanderish** [comment]
> @Wanderish Good luck setting it up! It's a whole new world of speed enabling GPU training.

Thanks much!
